In [ ]:
#| echo: false
# Run this to allow the use of R
# Cheekily, rpy2 doesn't work with the Matrix package of R,
# so I actually ran all the R using a document-wide R kernel,
# this is just for syntax highlighting...
%load_ext rpy2.ipython

Yesterday, we looked at a scRNA-seq dataset of <i style="color:#EB1960">Danio rerio</i> cells [@zebrafish-data].  We spent a lot of time understanding how the dataset was created, and ended with a bit of a mystery: <i style="color:#C0CF96">why did the cell counts not match up between our datasets?</i>  I don't have an answer to that mystery, unfortunately - but in my experience this mismatch happens in a lot of papers.  We might as well continue with the analysis, rather than getting hung up on a minor anomaly.

I realized since yesterday that there was another gene counts file, (the filtered tpms file).  I downloaded this and the "experiment metadata" files; I suspect these may be useful as I know that the TPM values were pre-quality-control:

> For the <b style="color:#EB1960">Smart-seq2 protocol</b> transcript per million (TPM) values reported by Salmon were used for the quality control (QC). Wells with fewer than 900 expressed genes (TPM > 1) or having more than either 60% of ERCC or 45% of mitochondrial content were annotated as <strong style="color:#C0CF96">poor quality cells</strong>. As a result, 322 cells failed QC and 542 single cells were selected for the further study.
>
> -- <cite><b style="color:#EB1960">Quality Control of Single-Cell Data; Materials and Methods Section;</b> @zebrafish-data </cite>

<details>
    <summary style="color:#C0CF96">
        <b>Can I mix R and Python in the same notebook?</b>
    </summary>
    Yes!  I sometimes use <b style="color:#EB1960">SOS Kernel</b> which allows me to swap between kernels at ease.  You can change kernels (R vs Python) using <span style="color:#757575">%use</span> magic commands; I've done this before but the syntax highlighting gets messed up often, so I've chosen not to. do that here.
</details>

In [ ]:
raw.counts <- Matrix::readMM('./localdata/E-MTAB-7117.expression_tpm.mtx')
dim(raw.counts)

[1] 21797   966

<details>
    <summary style="color:#C0CF96"><b>Does the metadata file help us resolve the cell count mystery?</b></summary>

What does this extra metadata file tell us?  Let's see:

In [ ]:
meta.data <- read.table("./localdata/E-MTAB-7117.sdrf.txt", sep='\t', header = TRUE)
head(meta.data)
dim(meta.data)
length(which(meta.data['Characteristics.single.cell.quality.'] != "not OK"))

Source.Name  Comment.ENA_SAMPLE. Comment.BioSD_SAMPLE.
1 CD4_gill_A1  ERS2634491          SAMEA4814592         
2 CD4_gill_A1  ERS2634491          SAMEA4814592         
3 CD4_gill_A10 ERS2634691          SAMEA4814793         
4 CD4_gill_A10 ERS2634691          SAMEA4814793         
5 CD4_gill_A11 ERS2635221          SAMEA4815323         
6 CD4_gill_A11 ERS2635221          SAMEA4815323         
  Characteristics.organism. Characteristics.strain. Characteristics.age.
1 Danio rerio               AB                      6                   
2 Danio rerio               AB                      6                   
3 Danio rerio               AB                      6                   
4 Danio rerio               AB                      6                   
5 Danio rerio               AB                      6                   
6 Danio rerio               AB                      6                   
  Unit.time.unit. Characteristics.developmental.stage. Characteristics.sex.
1 month           adult                                male                
2 month           adult                                male                
3 month           adult                                male                
4 month           adult                                male                
5 month           adult                                male                
6 month           adult                                male                
  Characteristics.genotype. ⋯ Comment.ENA_EXPERIMENT.
1 Tg(cd4-1:mCherry)         ⋯ ERX2737040             
2 Tg(cd4-1:mCherry)         ⋯ ERX2737040             
3 Tg(cd4-1:mCherry)         ⋯ ERX2737240             
4 Tg(cd4-1:mCherry)         ⋯ ERX2737240             
5 Tg(cd4-1:mCherry)         ⋯ ERX2737770             
6 Tg(cd4-1:mCherry)         ⋯ ERX2737770             
  Scan.Name                                  
1 SLX-10875.N701_S513.C9FTNANXX.s_5.r_1.fq.gz
2 SLX-10875.N701_S513.C9FTNANXX.s_5.r_2.fq.gz
3 SLX-10875.N712_S513.C9FTNANXX.s_5.r_1.fq.gz
4 SLX-10875.N712_S513.C9FTNANXX.s_5.r_2.fq.gz
5 SLX-10875.N714_S513.C9FTNANXX.s_5.r_1.fq.gz
6 SLX-10875.N714_S513.C9FTNANXX.s_5.r_2.fq.gz
  Comment.SUBMITTED_FILE_NAME.                Comment.ENA_RUN.
1 SLX-10875.N701_S513.C9FTNANXX.s_5.r_1.fq.gz ERR2723271      
2 SLX-10875.N701_S513.C9FTNANXX.s_5.r_2.fq.gz ERR2723271      
3 SLX-10875.N712_S513.C9FTNANXX.s_5.r_1.fq.gz ERR2723471      
4 SLX-10875.N712_S513.C9FTNANXX.s_5.r_2.fq.gz ERR2723471      
5 SLX-10875.N714_S513.C9FTNANXX.s_5.r_1.fq.gz ERR2724001      
6 SLX-10875.N714_S513.C9FTNANXX.s_5.r_2.fq.gz ERR2724001      
  Comment.FASTQ_URI.                                                            
1 ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR272/001/ERR2723271/ERR2723271_1.fastq.gz
2 ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR272/001/ERR2723271/ERR2723271_2.fastq.gz
3 ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR272/001/ERR2723471/ERR2723471_1.fastq.gz
4 ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR272/001/ERR2723471/ERR2723471_2.fastq.gz
5 ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR272/001/ERR2724001/ERR2724001_1.fastq.gz
6 ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR272/001/ERR2724001/ERR2724001_2.fastq.gz
  Comment.SPOT_LENGTH. Comment.READ_INDEX_1_BASE_COORD. Factor.Value.genotype.
1 250                  126                              Tg(cd4-1:mCherry)     
2 250                  126                              Tg(cd4-1:mCherry)     
3 250                  126                              Tg(cd4-1:mCherry)     
4 250                  126                              Tg(cd4-1:mCherry)     
5 250                  126                              Tg(cd4-1:mCherry)     
6 250                  126                              Tg(cd4-1:mCherry)     
  Factor.Value.organism.part. Factor.Value.single.cell.identifier.
1 gill                        CD4_gill_A1                         
2 gill                        CD4_gill_A1                         
3 gill                        CD4_gill_A10                        
4 gill                        C

[1] 2112   55

[1] 1952

At first glance, <b style="color:#C0CF96">no</b> - it does not.  The metadata has 2112 lines, and removing poor quality cells leaves us with 1952.

But, it does seem like some cells have multiple names, so let's filter by unique values.

In [ ]:
dim(unique(meta.data['Source.Name']))
length(
    unique(
        meta.data[
            which(meta.data['Characteristics.single.cell.quality.'] != "not OK"),
            "Source.Name"
        ]
    )
)

[1] 1056    1

[1] 976

Gee whiz, that 976 number returns...  We can re-use the list of 10 extra cells calculated yesterday to investigate them:

In [ ]:
discrepancies <- c(
    "ERR2723217", "ERR2723218", "ERR2723236", "ERR2723237", "ERR2723551",
    "ERR2723578", "ERR2723789", "ERR2723794", "ERR2723974", "ERR2723990"
)
extra.cells <- meta.data[
    which(sapply(meta.data["Comment.ENA_RUN."], `%in%`, discrepancies)),
]

In [ ]:
#| output: false
# Output hidden to reduce spam
extra.cells[1:10]
extra.cells[11:20]
extra.cells[21:30]
extra.cells[31:40]
extra.cells[41:55]

Source.Name           Comment.ENA_SAMPLE. Comment.BioSD_SAMPLE.
247  CD4_gut_D4            ERS2634771          SAMEA4814873         
248  CD4_gut_D4            ERS2634771          SAMEA4814873         
375  CD4_kidney_C5         ERS2634798          SAMEA4814900         
376  CD4_kidney_C5         ERS2634798          SAMEA4814900         
1053 LCK_gut_G8            ERS2634437          SAMEA4814538         
1054 LCK_gut_G8            ERS2634437          SAMEA4814538         
1055 LCK_gut_G9            ERS2634438          SAMEA4814539         
1056 LCK_gut_G9            ERS2634438          SAMEA4814539         
1305 LCK_thymus_B2         ERS2635210          SAMEA4815312         
1306 LCK_thymus_B2         ERS2635210          SAMEA4815312         
1333 LCK_thymus_C4         ERS2634456          SAMEA4814557         
1334 LCK_thymus_C4         ERS2634456          SAMEA4814557         
1335 LCK_thymus_C5         ERS2634457          SAMEA4814558         
1336 LCK_thymus_C5         ERS2634457          SAMEA4814558         
1545 mhc2dab_kidney_pl2_A2 ERS2635009          SAMEA4815111         
1546 mhc2dab_kidney_pl2_A2 ERS2635009          SAMEA4815111         
1609 mhc2dab_kidney_pl2_D1 ERS2635014          SAMEA4815116         
1610 mhc2dab_kidney_pl2_D1 ERS2635014          SAMEA4815116         
2099 Rag1_kidney_H3        ERS2635194          SAMEA4815296         
2100 Rag1_kidney_H3        ERS2635194          SAMEA4815296         
     Characteristics.organism. Characteristics.strain. Characteristics.age.
247  Danio rerio               AB                       6                  
248  Danio rerio               AB                       6                  
375  Danio rerio               AB                       6                  
376  Danio rerio               AB                       6                  
1053 Danio rerio               AB                       6                  
1054 Danio rerio               AB                       6                  
1055 Danio rerio               AB                       6                  
1056 Danio rerio               AB                       6                  
1305 Danio rerio               AB                       6                  
1306 Danio rerio               AB                       6                  
1333 Danio rerio               AB                       6                  
1334 Danio rerio               AB                       6                  
1335 Danio rerio               AB                       6                  
1336 Danio rerio               AB                       6                  
1545 Danio rerio               AB                      10                  
1546 Danio rerio               AB                      10                  
1609 Danio rerio               AB                      10                  
1610 Danio rerio               AB                      10                  
2099 Danio rerio               AB                       4                  
2100 Danio rerio               AB                       4                  
     Unit.time.unit. Characteristics.developmental.stage. Characteristics.sex.
247  month           adult                                male                
248  month           adult                                male                
375  month           adult                                male                
376  month           adult                                male                
1053 month           adult                                male                
1054 month           adult                                male                
1055 month           adult                                male                
1056 month           adult                                male                
1305 month           adult                                male                
1306 month           adult                                male                
1333 month           adult                                male                
1334 month  

Characteristics.organism.part. Characteristics.cell.type.
247  gut                            blood cell                
248  gut                            blood cell                
375  kidney                         blood cell                
376  kidney                         blood cell                
1053 gut                            blood cell                
1054 gut                            blood cell                
1055 gut                            blood cell                
1056 gut                            blood cell                
1305 thymus                         blood cell                
1306 thymus                         blood cell                
1333 thymus                         blood cell                
1334 thymus                         blood cell                
1335 thymus                         blood cell                
1336 thymus                         blood cell                
1545 kidney                         blood cell                
1546 kidney                         blood cell                
1609 kidney                         blood cell                
1610 kidney                         blood cell                
2099 kidney                         blood cell                
2100 kidney                         blood cell                
     Characteristics.phenotype.  Characteristics.individual.
247  mCherry positive cell       2                          
248  mCherry positive cell       2                          
375  mCherry positive cell       2                          
376  mCherry positive cell       2                          
1053 EGFP positive cell          1                          
1054 EGFP positive cell          1                          
1055 EGFP positive cell          1                          
1056 EGFP positive cell          1                          
1305 EGFP positive cell          1                          
1306 EGFP positive cell          1                          
1333 EGFP positive cell          1                          
1334 EGFP positive cell          1                          
1335 EGFP positive cell          1                          
1336 EGFP positive cell          1                          
1545 GFP and dsRed positive cell 3                          
1546 GFP and dsRed positive cell 3                          
1609 GFP and dsRed positive cell 3                          
1610 GFP and dsRed positive cell 3                          
2099 EGFP positive cell          4                          
2100 EGFP positive cell          4                          
     Characteristics.well.information. Characteristics.single.cell.quality.
247  single cell                       OK                                  
248  single cell                       OK                                  
375  single cell                       OK                                  
376  single cell                       OK                                  
1053 single cell                       OK                                  
1054 single cell                       OK                                  
1055 single cell                       OK                                  
1056 single cell                       OK                                  
1305 single cell                       OK                                  
1306 single cell                       OK                                  
1333 single cell                       OK                                  
1334 single cell                       OK                                  
1335 single cell                       OK                                  
1336 single cell                       OK                                  
1545 single cell                       OK                                  
1546 single cell                       OK                                  
1609 single cell                       OK                                  
1610 single cell                 

Protocol.REF Performer          Protocol.REF.1 Performer.1       
247  P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
248  P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
375  P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
376  P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1053 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1054 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1055 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1056 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1305 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1306 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1333 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1334 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1335 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1336 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1545 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1546 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1609 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
1610 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
2099 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
2100 P-MTAB-77662 Paulina Strzelecka P-MTAB-77663   Paulina Strzelecka
     Protocol.REF.2 Performer.2        Extract.Name          Material.Type.1
247  P-MTAB-77664   Paulina Strzelecka CD4_gut_D4            RNA            
248  P-MTAB-77664   Paulina Strzelecka CD4_gut_D4            RNA            
375  P-MTAB-77664   Paulina Strzelecka CD4_kidney_C5         RNA            
376  P-MTAB-77664   Paulina Strzelecka CD4_kidney_C5         RNA            
1053 P-MTAB-77664   Paulina Strzelecka LCK_gut_G8            RNA            
1054 P-MTAB-77664   Paulina Strzelecka LCK_gut_G8            RNA            
1055 P-MTAB-77664   Paulina Strzelecka LCK_gut_G9            RNA            
1056 P-MTAB-77664   Paulina Strzelecka LCK_gut_G9            RNA            
1305 P-MTAB-77664   Paulina Strzelecka LCK_thymus_B2         RNA            
1306 P-MTAB-77664   Paulina Strzelecka LCK_thymus_B2         RNA            
1333 P-MTAB-77664   Paulina Strzelecka LCK_thymus_C4         RNA            
1334 P-MTAB-77664   Paulina Strzelecka LCK_thymus_C4         RNA            
1335 P-MTAB-77664   Paulina Strzelecka LCK_thymus_C5         RNA            
1336 P-MTAB-77664   Paulina Strzelecka LCK_thymus_C5         RNA            
1545 P-MTAB-77664   Paulina Strzelecka mhc2dab_kidney_pl2_A2 RNA            
1546 P-MTAB-77664   Paulina Strzelecka mhc2dab_kidney_pl2_A2 RNA            
1609 P-MTAB-77664   Paulina Strzelecka mhc2dab_kidney_pl2_D1 RNA            
1610 P-MTAB-77664   Paulina Strzelecka mhc2dab_kidney_pl2_D1 RNA            
2099 P-MTAB-77664   Paulina Strzelecka Rag1_kidney_H3        RNA            
2100 P-MTAB-77664   Paulina Strzelecka Rag1_kidney_H3        RNA            
     Comment.single.cell.isolation. Comment.library.construction.
247  FACS                           smart-seq2                   
248  FACS                           smart-seq2                   
375  FACS                           smart-seq2                   
376  FACS                           smart-seq2                   
1053 FACS                           smart-seq2                   
1054 FACS                           smart-seq2                   
1055 FACS                           smart-seq2                   
1056 FACS                           smart-seq2                   
1305 FACS                           smart-seq2                   
1306 FACS                           smart-seq2                   
1333 FACS                           smart-seq2                   
1334 FACS                           smart-seq2                   
1335 FACS                           sma

Comment.input.molecule. Comment.primer. Comment.end.bias.
247  polyA RNA               oligo-dT        none             
248  polyA RNA               oligo-dT        none             
375  polyA RNA               oligo-dT        none             
376  polyA RNA               oligo-dT        none             
1053 polyA RNA               oligo-dT        none             
1054 polyA RNA               oligo-dT        none             
1055 polyA RNA               oligo-dT        none             
1056 polyA RNA               oligo-dT        none             
1305 polyA RNA               oligo-dT        none             
1306 polyA RNA               oligo-dT        none             
1333 polyA RNA               oligo-dT        none             
1334 polyA RNA               oligo-dT        none             
1335 polyA RNA               oligo-dT        none             
1336 polyA RNA               oligo-dT        none             
1545 polyA RNA               oligo-dT        none             
1546 polyA RNA               oligo-dT        none             
1609 polyA RNA               oligo-dT        none             
1610 polyA RNA               oligo-dT        none             
2099 polyA RNA               oligo-dT        none             
2100 polyA RNA               oligo-dT        none             
     Comment.LIBRARY_LAYOUT. Comment.LIBRARY_SELECTION.
247  PAIRED                  Oligo-dT                  
248  PAIRED                  Oligo-dT                  
375  PAIRED                  Oligo-dT                  
376  PAIRED                  Oligo-dT                  
1053 PAIRED                  Oligo-dT                  
1054 PAIRED                  Oligo-dT                  
1055 PAIRED                  Oligo-dT                  
1056 PAIRED                  Oligo-dT                  
1305 PAIRED                  Oligo-dT                  
1306 PAIRED                  Oligo-dT                  
1333 PAIRED                  Oligo-dT                  
1334 PAIRED                  Oligo-dT                  
1335 PAIRED                  Oligo-dT                  
1336 PAIRED                  Oligo-dT                  
1545 PAIRED                  Oligo-dT                  
1546 PAIRED                  Oligo-dT                  
1609 PAIRED                  Oligo-dT                  
1610 PAIRED                  Oligo-dT                  
2099 PAIRED                  Oligo-dT                  
2100 PAIRED                  Oligo-dT                  
     Comment.LIBRARY_SOURCE.    Comment.LIBRARY_STRAND.
247  TRANSCRIPTOMIC SINGLE CELL not applicable         
248  TRANSCRIPTOMIC SINGLE CELL not applicable         
375  TRANSCRIPTOMIC SINGLE CELL not applicable         
376  TRANSCRIPTOMIC SINGLE CELL not applicable         
1053 TRANSCRIPTOMIC SINGLE CELL not applicable         
1054 TRANSCRIPTOMIC SINGLE CELL not applicable         
1055 TRANSCRIPTOMIC SINGLE CELL not applicable         
1056 TRANSCRIPTOMIC SINGLE CELL not applicable         
1305 TRANSCRIPTOMIC SINGLE CELL not applicable         
1306 TRANSCRIPTOMIC SINGLE CELL not applicable         
1333 TRANSCRIPTOMIC SINGLE CELL not applicable         
1334 TRANSCRIPTOMIC SINGLE CELL not applicable         
1335 TRANSCRIPTOMIC SINGLE CELL not applicable         
1336 TRANSCRIPTOMIC SINGLE CELL not applicable         
1545 TRANSCRIPTOMIC SINGLE CELL not applicable         
1546 TRANSCRIPTOMIC SINGLE CELL not applicable         
1609 TRANSCRIPTOMIC SINGLE CELL not applicable         
1610 TRANSCRIPTOMIC SINGLE CELL not applicable         
2099 TRANSCRIPTOMIC SINGLE CELL not applicable         
2100 TRANSCRIPTOMIC SINGLE CELL not applicable         
     Comment.LIBRARY_STRATEGY. Comment.NOMINAL_LENGTH. Comment.NOMINAL_SDEV.
247  RNA-Seq                   400                     20                   
248  RNA-Seq                   400                     20                   
375  RNA-Seq                   400                     20                   
376  RNA-Seq          

Comment.ORIENTATION. Protocol.REF.3
247  5-3-5-3              P-MTAB-77665  
248  5-3-5-3              P-MTAB-77665  
375  5-3-5-3              P-MTAB-77665  
376  5-3-5-3              P-MTAB-77665  
1053 5-3-5-3              P-MTAB-77665  
1054 5-3-5-3              P-MTAB-77665  
1055 5-3-5-3              P-MTAB-77665  
1056 5-3-5-3              P-MTAB-77665  
1305 5-3-5-3              P-MTAB-77665  
1306 5-3-5-3              P-MTAB-77665  
1333 5-3-5-3              P-MTAB-77665  
1334 5-3-5-3              P-MTAB-77665  
1335 5-3-5-3              P-MTAB-77665  
1336 5-3-5-3              P-MTAB-77665  
1545 5-3-5-3              P-MTAB-77665  
1546 5-3-5-3              P-MTAB-77665  
1609 5-3-5-3              P-MTAB-77665  
1610 5-3-5-3              P-MTAB-77665  
2099 5-3-5-3              P-MTAB-77665  
2100 5-3-5-3              P-MTAB-77665  
     Performer.3                                Assay.Name           
247  Cambridge Institute Genomics Core Facility CD4_gut_D4           
248  Cambridge Institute Genomics Core Facility CD4_gut_D4           
375  Cambridge Institute Genomics Core Facility CD4_kidney_C5        
376  Cambridge Institute Genomics Core Facility CD4_kidney_C5        
1053 Cambridge Institute Genomics Core Facility LCK_gut_G8           
1054 Cambridge Institute Genomics Core Facility LCK_gut_G8           
1055 Cambridge Institute Genomics Core Facility LCK_gut_G9           
1056 Cambridge Institute Genomics Core Facility LCK_gut_G9           
1305 Cambridge Institute Genomics Core Facility LCK_thymus_B2        
1306 Cambridge Institute Genomics Core Facility LCK_thymus_B2        
1333 Cambridge Institute Genomics Core Facility LCK_thymus_C4        
1334 Cambridge Institute Genomics Core Facility LCK_thymus_C4        
1335 Cambridge Institute Genomics Core Facility LCK_thymus_C5        
1336 Cambridge Institute Genomics Core Facility LCK_thymus_C5        
1545 Cambridge Institute Genomics Core Facility mhc2dab_kidney_pl2_A2
1546 Cambridge Institute Genomics Core Facility mhc2dab_kidney_pl2_A2
1609 Cambridge Institute Genomics Core Facility mhc2dab_kidney_pl2_D1
1610 Cambridge Institute Genomics Core Facility mhc2dab_kidney_pl2_D1
2099 Cambridge Institute Genomics Core Facility Rag1_kidney_H3       
2100 Cambridge Institute Genomics Core Facility Rag1_kidney_H3       
     Technology.Type  Comment.ENA_EXPERIMENT.
247  sequencing assay ERX2737320             
248  sequencing assay ERX2737320             
375  sequencing assay ERX2737347             
376  sequencing assay ERX2737347             
1053 sequencing assay ERX2736986             
1054 sequencing assay ERX2736986             
1055 sequencing assay ERX2736987             
1056 sequencing assay ERX2736987             
1305 sequencing assay ERX2737759             
1306 sequencing assay ERX2737759             
1333 sequencing assay ERX2737005             
1334 sequencing assay ERX2737005             
1335 sequencing assay ERX2737006             
1336 sequencing assay ERX2737006             
1545 sequencing assay ERX2737558             
1546 sequencing assay ERX2737558             
1609 sequencing assay ERX2737563             
1610 sequencing assay ERX2737563             
2099 sequencing assay ERX2737743             
2100 sequencing assay ERX2737743             
     Scan.Name                                  
247  SLX-12114.i720_i506.HKKG2BBXX.s_1.r_1.fq.gz
248  SLX-12114.i720_i506.HKKG2BBXX.s_1.r_2.fq.gz
375  SLX-10874.N705_S505.C9FTNANXX.s_4.r_1.fq.gz
376  SLX-10874.N705_S505.C9FTNANXX.s_4.r_2.fq.gz
1053 SLX-12119.i710_i521.HKCTNBBXX.s_1.r_1.fq.gz
1054 SLX-12119.i710_i521.HKCTNBBXX.s_1.r_2.fq.gz
1055 SLX-12119.i711_i521.HKCTNBBXX.s_1.r_1.fq.gz
1056 SLX-12119.i711_i521.HKCTNBBXX.s_1.r_2.fq.gz
1305 SLX-12119.i702_i503.HKCTNBBXX.s_1.r_1.fq.gz
1306 SLX-12119.i702_i503.HKCTNBBXX.s_1.r_2.fq.gz
1333 SLX-12119.i704_i505.HKCTNBBXX.s_1.r_1.fq.gz
1334 SLX-12119.i704_i505.HKCTNBBXX.s_1.r_2.fq.gz
1335 SLX-12119.i705_i505.HKCTNBBXX.s_1.r_1.fq.gz
1336 SLX-12119.i705_i5

But as before there doesn't really seem to be any outstanding features.  So the mystery continues!

</details>

# Quality Control

The <b style="color:#757575">raw.counts</b> matrix is a <b style="color:#A6A440">Sparse Matrix</b> with 21,797 genes and 966 cells - the values are expressed in TPM.  Our goal is to filter out the low quality cells.

> For the <b style="color:#EB1960">Smart-seq2 protocol</b> transcript per million (TPM) values reported by Salmon were used for the quality control (QC). Wells with fewer than 900 expressed genes (TPM > 1) or having more than either 60% of ERCC or 45% of mitochondrial content were annotated as <strong style="color:#C0CF96">poor quality cells</strong>. As a result, 322 cells failed QC and 542 single cells were selected for the further study.
>
> -- <cite><b style="color:#EB1960">Quality Control of Single-Cell Data; Materials and Methods Section;</b> @zebrafish-data </cite>

At the time, I didn't think we could find the mitochondrial genes with the data we had (and same with ERCC).  I've gone back and edited this though, in the dropdowns.

<details>
    <summary style="color:#C0CF96"><b>Getting mitochondrial genes</b></summary>

I used Python since chronologically I did this ater I gave up on R.  The work to create the `mito-genes` csv was done in the blog post [Exploring Ensembl](./004_005_ensemble.html).

In [ ]:
mito_genes = pd.read_csv("./localdata/mito-genes.csv")
mito_genes["gene_id"].head()

0    ENSDARG00000083480
1    ENSDARG00000082753
2    ENSDARG00000081443
3    ENSDARG00000080337
4    ENSDARG00000083046
Name: gene_id, dtype: object


Now we want to calculate the percent of genomic material per cell that is contained in just the genes in `mito_genes`, which should not be too hard!  Left as excersize to the reader.

</details>

<details>
    <summary style="color:#C0CF96"><b>Getting ERCC genes</b></summary>

I could not figure out how to do this; I think they have already been removed from the dataset that we have.

</details>

But yesterday we noticed that we could get the 542 futher-study cells by looking at the metadata and only grabbing the cells which were eventually assigned a cluster by the researchers.

In [ ]:
# Calculate the cells we want to study
sample.data <- read.table("./localdata/ExpDesign-E-MTAB-7117.tsv", sep='\t', header=TRUE)
filtered <- sample.data[which(
    sample.data['Sample.Characteristic.cluster.'] != "unknown"
),]
cells.to.study <- filtered$Assay
length(cells.to.study)

[1] 542

In [ ]:
# Get which rows correspond to which assay (cell)
assay.to.row.map <- read.table(
    './localdata/E-MTAB-7117.expression_tpm.mtx_cols'
)$V1
length(assay.to.row.map)

[1] 966

In [ ]:
# Subset the raw.counts matrix
raw.counts.filtered <- raw.counts[,
    which(sapply(assay.to.row.map, `%in%`, cells.to.study)),
]
dim(raw.counts.filtered)
# Save the matrix, for posterity
writeMM(raw.counts.filtered, "./localdata/542_cells_21797_genes.tpm.mtx")

[1] 21797   542

NULL

Now we need to perform quality control on the genes;

> For each of the <strong style="color:#C0CF96">542 single cells</strong>, counts reported by <b style="color:#EB1960">Salmon</b> were transformed into normalised counts per million (CPM) and used for the further analysis. This was performed by <strong style="color:#C0CF96">dividing the number of counts for each gene with the total number of counts for each cell and by multiplying the resulting number by a factor of 1,000,000</strong>. Genes that were expressed in less than 1% of cells (e.g. 5 single cells with CPM > 1) <strong style="color:#C0CF96">were filtered out</strong>. In the final step we ended up using 16,059 genes across the 542 single cells. The <b style="color:#EB1960">scran R package</b> (version 1.6.7) @scran was then used to <b style="color:#537FBF">normalise</b> the data and remove differences due to the library size or capture efficiency and sequencing depth.
>
> -- <cite><b style="color:#EB1960">Downstream Analysis of Smart-seq2 Data; Materials and Methods Section;</b> @zebrafish-data </cite>

It was harder than expected, as I am less-than-fluent with R and their sparse matrix operations weren't as analogous to dense matrices as I would have liked.  First let's take a look at our the sparse datatype we have.  I found the blogpost by @dgTMatrix useful, but ultimately got fed up with R and decided to transition to python.

<details><summary><span class="false-starts" style="font-weight:bold;">False Start</span>: <b style="color:#C0CF96">Trying to use R</b></summary>

In [ ]:
str(raw.counts.filtered)

Formal class 'dgTMatrix' [package "Matrix"] with 6 slots
  ..@ i       : int [1:1156509] 4 10 19 36 56 57 58 59 62 69 ...
  ..@ j       : int [1:1156509] 0 0 0 0 0 0 0 0 0 0 ...
  ..@ Dim     : int [1:2] 21797 542
  ..@ Dimnames:List of 2
  .. ..$ : NULL
  .. ..$ : NULL
  ..@ x       : num [1:1156509] 0.593 145.541 361.733 0.178 0.107 ...
  ..@ factors : list()


`(i, j)` gives us the rows and columns, with the value being the corresponding entry in `x`.  The `T` in `dgTMatrix` stands for triplet, because it's essentially just a list of triplets `(i, j, x)`.  Another common format is `CsparseMatrix`, although the explanation is more complicated.  It's explained well by @dgTMatrix, I'm just in a rush to get to actual data analysis!

In [ ]:
raw.counts.csparse <- str(as(raw.counts.filtered, "CsparseMatrix"))

Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  ..@ i       : int [1:1156509] 4 10 19 36 56 57 58 59 62 69 ...
  ..@ p       : int [1:543] 0 1610 3532 5491 7585 9273 11182 13157 14623 15552 ...
  ..@ Dim     : int [1:2] 21797 542
  ..@ Dimnames:List of 2
  .. ..$ : NULL
  .. ..$ : NULL
  ..@ x       : num [1:1156509] 0.593 145.541 361.733 0.178 0.107 ...
  ..@ factors : list()


Now, we want to transform the data into <b style="color:#C0CF96">normalized counts per million</b>.

In [ ]:
# This will throw an error
colSums(raw.counts.csparse)

ERROR: Error in colSums(raw.counts.csparse): 'x' must be an array of at least two dimensions


<details><summary style="color:#C0CF96;font-weight:bold;">But we know it should work for sparse matrixes from the docs</summary>

![What happens when we run `?colSums` to get the docs](r-help.png)

Fun fact: for some reason running the R help command will mess with Quarto's html, breaking your webpage!

</details>


In [ ]:
# We can try to force it to use the Matrix overload, to no avail.
Matrix::colSums(raw.counts.csparse)

ERROR: Error in base::rowSums(x, na.rm = na.rm, dims = dims, ...): 'x' must be an array of at least two dimensions


And it's about here that I get fed up with R.

</details>

Let's load our `raw.counts.filtered` matrix in python:

In [ ]:
from scipy.io import mmread
from scipy import sparse
import numpy as np
import pandas as pd

In [ ]:
raw_counts_filtered = sparse.csc_array(mmread(
    './localdata/542_cells_21797_genes.tpm.mtx'
))
raw_counts_filtered

<21797x542 sparse array of type '<class 'numpy.float64'>'
	with 1156509 stored elements in Compressed Sparse Column format>

And now we'll copy the process they used to rescale their data in terms of genes per cell.

In [ ]:
genes_per_cell = raw_counts_filtered.sum(axis=0)
raw_counts_nonzero = raw_counts_filtered[:, genes_per_cell > 0]
genes_per_cell = genes_per_cell[genes_per_cell > 0]
raw_counts_nonzero = raw_counts_nonzero / genes_per_cell * 1e6
raw_counts_normed = sparse.csr_array(raw_counts_nonzero)
raw_counts_normed

<21797x542 sparse array of type '<class 'numpy.float64'>'
	with 1156509 stored elements in Compressed Sparse Row format>

Now we need to filter out genes who were expressed in less than 1% of cells.  They're a bit more strict than that - it has to have a CPM of at least 1 to count as being "expressed".

In [ ]:
raw_counts_expressed = raw_counts_normed.copy()
raw_counts_expressed[raw_counts_expressed <= 1] = 0
raw_counts_expressed[raw_counts_expressed > 1] = 1
number_of_cells_gene_appears_in = raw_counts_expressed.sum(axis=1)
raw_counts_genes_filtered = raw_counts_normed[
    number_of_cells_gene_appears_in > 5,
    :
].copy()

In [ ]:
raw_counts_genes_filtered

<17832x542 sparse array of type '<class 'numpy.float64'>'
	with 1141288 stored elements in Compressed Sparse Row format>

We have too many genes and I'm not sure why - however, I can't find anywhere in their paper a count of the pre-filtered genes.  Just like how their dataset contains more cells than reported, it may contain more genes than reported!

# Analysis

> In order to identify the <b style="color:#A6A440">highly variable genes (HVGs)</b> we utilised the <b style="color:#537FBF">Brennecke Method</b> [@brennecke]. We inferred the noise model from the <b style="color:#A6A440">ERCCs</b> and <b style="color:#C0CF96">selected genes that vary higher than 20% percentage of variation</b>. This was performed by using the
“<b style="color:#757575">BrenneckeGetVariableGenes</b>” command of <b style="color:#EB1960">M3Drop v1.4.0 R package</b> setting fdr equal to 0.01 and minimum percentage of variance due to biological factors (minBiolDisp) equal to 0.2. <b style="color:#C0CF96">In total, 3,374 were annotated as HVGs.</b>
>
> -- <cite><b style="color:#EB1960">Downstream Analysis of Smart-seq2 Data; Materials and Methods Section;</b> @zebrafish-data </cite>

The `BrenneckeGetVariableGenes` method is described in the <b style="color:#EB1960">M3Drop documentation</b> as follows:

![`BrenneckeGetVariableGenes`](./m3drop.png)

Thus we can see that we need the ERCCs to deduce the HGVs.  However, I can't find any data on the ERCC spike-ins, so at this point we stop here.  Sorry to disappoint.

### References

::: {#refs}
:::